### PR TITLE
SQL-1279: add setup gui for dsn

### PR DESCRIPTION
**Pull request by terakilobyte**
I think this is nearly there. I've a few outstanding questions about what we should do with a few things that I want to wrap up before marking it as ready for merge.

### CODE REVIEW

File: win_setupgui/src/gui/config.rs
Comment (terakilobyte) [inline win_setupgui/src/gui/config.rs:68]: I still need to make this a file dialogue with directory-picking behavior.

File: core/src/odbc_uri.rs
Comment (pmeredit) [reply core/src/odbc_uri.rs:21]: Might as well!

File: core/src/odbc_uri.rs
Comment (pmeredit) [inline core/src/odbc_uri.rs:47]: Let's make this `pub(crate)`. I don't want to leak the abstraction outside of the core crate.

File: odbc/src/api/mod.rs
Comment (pmeredit) [reply odbc/src/api/mod.rs:11]: Yea, we can call it whatever

File: core/src/odbc_uri.rs
Comment (pmeredit) [reply core/src/odbc_uri.rs:47]: Actually, it looks like all you are using this for is to iterate over the keys. I'd just add an iter() method for ODBCUri that calls 0.iter and not make this `pub(crate)` either.

**Comment by terakilobyte** [comment]
@pmeredit @nbagnard With the proposed changes here, namely the inclusion of the `logpath`, I'd like to create a ticket to use that (if available) by the `file_dbg_macros`. Is it okay to create that ticket greedily while this is in review?

File: core/src/util/registry.rs
Comment (pmeredit) [inline core/src/util/registry.rs:1]: This cfg here is redundant with the one in mod.rs. I would say just leave the one in mod.rs (and presumably we'll have a module called ini for the other platforms).

File: core/src/util/registry.rs
Comment (pmeredit) [inline core/src/util/registry.rs:2]: I don't think we need this module here, unless we want to change this to have non windows as a module in this file. In that case we'd want to take the cfg out of the mod file above. I'm fine with going either direction, but let's choose one.

File: core/src/util/registry.rs
Comment (pmeredit) [inline core/src/util/registry.rs:73]: Since we are actually returning errors, I think it makes more sense to format the errors with this message and only call `dbg_write` at the top where the error last manifests. Reduces redundant code. Do this in any function that returns `Result`. Obviously, the ones that don't return `Result` should call dbg_write :)

File: cstr/src/lib.rs
Comment (pmeredit) [inline cstr/src/lib.rs:1]: This is a great change.

File: cstr/src/lib.rs
Comment (pmeredit) [inline cstr/src/lib.rs:50]: For consistency, let's called this `input_atext_to_string`

Comment (pmeredit) [reply cstr/src/lib.rs:50]: Or we can move `a` and `w` as suffixes. Either way, we should make it clear that this is for ansi text.

File: win_setupgui/src/api/functions.rs
Comment (pmeredit) [reply win_setupgui/src/api/functions.rs:12]: @nbagnard the length is defined as 32 in the header, the tokens themselves are from the spec

**Review by pmeredit — CHANGES_REQUESTED** [review]
Looks great, just a few changes.

File: core/src/util/dsn.rs
Comment (nbagnard) [inline core/src/util/dsn.rs:64]: I learned something today and I'll let you decide if this needs to be updated or not.
The value associated to the DSN name under datasource is user-defined so we can do what we want. [The recommendation is to use the name of the associated DBMS](https://learn.microsoft.com/en-us/sql/odbc/reference/install/odbc-data-sources-subkey) (DBMS_NAME from constant in our case would work).

File: core/src/util/registry.rs
Comment (nbagnard) [inline core/src/util/registry.rs:6]: I am now unsure if we can by-pass calling SQLWriteDSNToIni and SQLWritePrivateProfileString because the root key is different depending on the type of DSN created :'(
It is HKEY_LOCAL_MACHINE for System DSN and HKEY_CURRENT_USER for user DSN. 
SQLGetConfigMode could tell us : https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlgetconfigmode-function
SQLConfigDataSource does not pass the info to ConfigDSN but maps it to User DSN .... https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlconfigdatasource-function?view=sql-server-ver16#comments

File: core/src/util/registry.rs
Comment (nbagnard) [inline core/src/util/registry.rs:12]: I think that add_dsn and add_to_datasource go hand in hand. Same goes for remove_dsn and remove_from_datasource. One should be calling the other or the code should be in only one place, otherwise you might end up with a partial DSN config, no?

File: integration_test/tests/common/mod.rs
Comment (nbagnard) [inline integration_test/tests/common/mod.rs:17]: Why this change? The name I see in our reg setup script is `MongoDB Atlas SQL ODBC Driver` and Power-BI is using this one as well.
I don't know if driver names are case-sensitive, so maybe it does not matter. Still, I am curious about it ?

**Review by nbagnard — CHANGES_REQUESTED** [review]
Sorry, this is a partial review. I did not look into the UI specific part in details yet, but I think it is important that we figure out the User vs system DSN part so I'd rather give my feedback now instead of waiting another day.

File: win_setupgui/src/gui/config.rs
Comment (nbagnard) [inline win_setupgui/src/gui/config.rs:34]:
```suggestion
    #[nwg_control(flags: "VISIBLE", text: "User name")]
```

File: win_setupgui/src/gui/config.rs
Comment (nbagnard) [inline win_setupgui/src/gui/config.rs:1]: Could we hide the 'Test' button and 'LogPath' file picker until it can be hooked to something?

Comment (terakilobyte) [reply win_setupgui/src/gui/config.rs:1]: Absolutely. It's as easy as commenting them out, mostly.

File: win_setupgui/src/gui/config.rs
Comment (terakilobyte) [reply win_setupgui/src/gui/config.rs:188]: I did. I thought it would be nice to have this in a log.

File: integration_test/tests/common/mod.rs
Comment (terakilobyte) [reply integration_test/tests/common/mod.rs:17]: Oh they absolutely are. This was probably a fat finger vim uppercase. Good catch.

File: core/src/util/registry.rs
Comment (terakilobyte) [reply core/src/util/registry.rs:12]: Ah, I segmented them because, in my mind, the datasource is under `ODBC.INI\ODBC Data Sources`, whereas dsn is under `ODBC.INI`. This is largely gone now though since I've discovered what the installer dll is actually called (odbccp32.dll) and am using those functions.

File: core/src/util/registry.rs
Comment (terakilobyte) [reply core/src/util/registry.rs:6]: SQLGetConfigMode is queryable and tells us the mode. We may end up having to go this route, as using SQLWritePrivateProfileString and SQLGetPrivateProfileString are extremely... flaky. This is something easier to show synchronously, but they have unpredictable behavior it seems.

File: core/src/odbc_uri.rs
Comment (terakilobyte) [reply core/src/odbc_uri.rs:22]: Absolutely. I can  also disable them show they aren't shown to the user, and skip their logic. If that works for you, I'll ensure to add that ticket number in comments.

File: win_setupgui/src/gui/config.rs
Comment (terakilobyte) [reply win_setupgui/src/gui/config.rs:1]: I've commented out the logpath stuff and added SQL-1281 as the reason.

For the test button, I thought we wanted that since we won't ship without test functionality.

File: core/src/util/mod.rs
Comment (terakilobyte) [inline core/src/util/mod.rs:8]: Marking these two as windows only for now. When we support macos/linux we'll need to figure out which library to link against for the function we need, and dsn can be agnostic.

File: core/src/util/dsn.rs
Comment (nbagnard) [inline core/src/util/dsn.rs:179]: [Discussed offline] Let's do a 1 to 1 mapping between DSNOpt and ODBCuri.

**Review by nbagnard — CHANGES_REQUESTED** [review]
A couple of last small things that I noticed, but overall this is looking very good!!!

File: core/src/util/dsn.rs
Comment (terakilobyte) [reply core/src/util/dsn.rs:77]: Sounds good. I'll add a comment!

File: core/src/util/dsn.rs
Comment (terakilobyte) [reply core/src/util/dsn.rs:126]: I will add a hard check. I can't safeguard against it, but it should be impossible based on Microsoft's registry value documentation. They could change it in the future, which could cause this to fail. 🤔 

This may be a good place for a debug write, with something like "If you see this error please report it."

File: win_setupgui/Cargo.toml
Comment (terakilobyte) [reply win_setupgui/Cargo.toml:17]: native-windows-* is needed for the ui iself. I'll follow up on odbc-sys and remove if it's safe.

File: win_setupgui/src/config.rs
Comment (terakilobyte) [reply win_setupgui/src/config.rs:15]: I'm not sure if ODBC Admin uses this info, but I'll look into it.

File: win_setupgui/src/config.rs
Comment (nbagnard) [reply win_setupgui/src/config.rs:49]: The logic is already geared toward a quirk from DSN options being passed, isn't it?
Maybe just update the function description to detail that it provides a `;` separated string, and we should be all set.

File: core/src/util/dsn.rs
Comment (pmeredit) [inline core/src/util/dsn.rs:61]: This giant tuple match feels weird. Might as well make it a vec and do
```
if v.iter().all(|x|x) {
    Ok(...)
} else if v[1] == false {
    Err(...)
} else {
    Err(other)
}
```

File: win_setupgui/src/gui.rs
Comment (pmeredit) [inline win_setupgui/src/gui.rs:72]: Some of these commented out fields appear to be the same. Let's get rid of the duplicates.

**Review by pmeredit — CHANGES_REQUESTED** [review]
Looks pretty good! Just two changes

File: win_setupgui/src/gui.rs
Comment (terakilobyte) [reply win_setupgui/src/gui.rs:72]: I looked again and I didn't see any duplicates of the same name. The button in this line is required to actually pop the dialog open.

Comment (pmeredit) [reply win_setupgui/src/gui.rs:72]: Ah, ok

**Review by pmeredit — APPROVED** [review]
LGTM

File: core/src/util/dsn.rs
Comment (nbagnard) [reply core/src/util/dsn.rs:139]: Yes, I think a debug write with the key name could be useful. But unless they have a bunch of giant values, it should easy to figure out which key is problematic

File: win_setupgui/Cargo.toml
Comment (nbagnard) [reply win_setupgui/Cargo.toml:17]: ok, so looks like it's needed after all :)

File: win_setupgui/src/config.rs
Comment (nbagnard) [reply win_setupgui/src/config.rs:22]: [Discussed offline] ODBC admin does not seem to read posted errors. So we won't post at least for now.

File: cstr/src/lib.rs
Comment (nbagnard) [inline cstr/src/lib.rs:20]:
```suggestion
/// It assumes null termination if the supplied length is negative.
```

File: cstr/src/lib.rs
Comment (nbagnard) [inline cstr/src/lib.rs:47]:
```suggestion
/// It assumes null termination if the supplied length is negative.
```

**Review by nbagnard — APPROVED** [review]
LGTM! 🎉 💯

**Comment by terakilobyte** [comment]
evergreen merge

**Pull request merged**
